File: tensorflow_io/python/ops/bigtable/bigtable_row_range.py
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Module implementing basic functions for obtaining BigTable RowRanges"""

from tensorflow_io.python.ops import core_ops


class RowRange:
    """Object representing a range of rows."""

    def __init__(self, impl) -> None:
        self._impl = impl

    def __repr__(self) -> str:
        print(self._impl)
        return core_ops.bigtable_print_row_range(self._impl).numpy()[0].decode()


def infinite() -> RowRange:
    """Create a infinite row range.
    Returns:
      RowRange: Infinite RowRange."""
    return RowRange(core_ops.bigtable_row_range("", False, "", False))


def starting_at(row_key: str) -> RowRange:
    """Create a row range staring at given row (inclusive).

    Args:
      row_key (str): The starting row key of the range (inclusive).
    Returns:
      RowRange: The row range which starts at `row_key` (inclusive).
    """
    return RowRange(core_ops.bigtable_row_range(row_key, False, "", False))


def ending_at(row_key: str) -> RowRange:
    """Create a row range ending at given row (inclusive).

    Args:
      row_key (str): The ending row key of the range (inclusive).
    Returns:
      RowRange: The row range which ends at `row_key` (inclusive).
    """
    return RowRange(core_ops.bigtable_row_range("", False, row_key, False))


def empty() -> RowRange:
    """Create an empty row range.
    Returns:
      RowRange: Empty RowRange."""
    return RowRange(core_ops.bigtable_empty_row_range())


def prefix(prefix_str: str) -> RowRange:
    """Create a range of all rows starting with a given prefix.

    Args:
      prefix_str (str): The prefix with which all rows start
    Returns:
      RowRange: The row range of all rows starting with the given prefix.
    """
    return RowRange(core_ops.bigtable_prefix_row_range(prefix_str))


def right_open(start: str, end: str) -> RowRange:
    """Create a row range exclusive at the start and inclusive at the end.

    Args:
      start (str): The start of the row range (inclusive).
      end (str): The end of the row range (exclusive).
    Returns:
      RowRange: The row range between the `start` and `end`.
    """
    return RowRange(core_ops.bigtable_row_range(start, False, end, True))


def left_open(start: str, end: str) -> RowRange:
    """Create a row range inclusive at the start and exclusive at the end.

    Args:
      start (str): The start of the row range (exclusive).
      end (str): The end of the row range (inclusive).
    Returns:
      RowRange: The row range between the `start` and `end`.
    """
    return RowRange(core_ops.bigtable_row_range(start, True, end, False))


def open_range(start: str, end: str) -> RowRange:
    """Create a row range exclusive at both the start and the end.

    Args:
      start (str): The start of the row range (exclusive).
      end (str): The end of the row range (exclusive).
    Returns:
      RowRange: The row range between the `start` and `end`.
    """
    return RowRange(core_ops.bigtable_row_range(start, True, end, True))


def closed_range(start: str, end: str) -> RowRange:
    """Create a row range inclusive at both the start and the end.

    Args:
      start (str): The start of the row range (inclusive).
      end (str): The end of the row range (inclusive).
    Returns:
      RowRange: The row range between the `start` and `end`.
    """
    return RowRange(core_ops.bigtable_row_range(start, False, end, False))

File: tensorflow_io/core/kernels/bigtable/bigtable_row_range.cc
/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/
#include "tensorflow_io/core/kernels/bigtable/bigtable_row_range.h"

namespace cbt = ::google::cloud::bigtable;

namespace tensorflow {
namespace io {

class BigtableEmptyRowRangeOp
    : public AbstractBigtableResourceOp<BigtableRowRangeResource> {
 public:
  explicit BigtableEmptyRowRangeOp(OpKernelConstruction* ctx)
      : AbstractBigtableResourceOp<BigtableRowRangeResource>(ctx) {
    VLOG(1) << "BigtableEmptyRowRangeOp ctor ";
  }

 private:
  StatusOr<BigtableRowRangeResource*> CreateResource() override {
    return new BigtableRowRangeResource(cbt::RowRange::Empty());
  }
};

REGISTER_KERNEL_BUILDER(Name("BigtableEmptyRowRange").Device(DEVICE_CPU),
                        BigtableEmptyRowRangeOp);

class BigtableRowRangeOp
    : public AbstractBigtableResourceOp<BigtableRowRangeResource> {
 public:
  explicit BigtableRowRangeOp(OpKernelConstruction* ctx)
      : AbstractBigtableResourceOp<BigtableRowRangeResource>(ctx) {
    VLOG(1) << "BigtableRowRangeOp ctor ";
    OP_REQUIRES_OK(ctx, ctx->GetAttr("left_row_key", &left_row_key_));
    OP_REQUIRES_OK(ctx, ctx->GetAttr("left_open", &left_open_));
    OP_REQUIRES_OK(ctx, ctx->GetAttr("right_row_key", &right_row_key_));
    OP_REQUIRES_OK(ctx, ctx->GetAttr("right_open", &right_open_));
  }

 private:
  StatusOr<BigtableRowRangeResource*> CreateResource() override {
    VLOG(1) << "BigtableRowRangeOp constructing row_range:"
            << (left_open_ ? "(" : "[") << left_row_key_ << ":"
            << right_row_key_ << (right_open_ ? ")" : "]");

    // both empty - infinite
    if (left_row_key_.empty() && right_row_key_.empty()) {
      return new BigtableRowRangeResource(cbt::RowRange::InfiniteRange());
    }

    // open
    if (left_open_ && right_open_) {
      return new BigtableRowRangeResource(
          cbt::RowRange::Open(left_row_key_, right_row_key_));
    }
    // closed
    if (!left_open_ && !right_open_) {
      return new BigtableRowRangeResource(
          cbt::RowRange::Closed(left_row_key_, right_row_key_));
    }
    // right_open
    if (!left_open_ && right_open_) {
      return new BigtableRowRangeResource(
          cbt::RowRange::RightOpen(left_row_key_, right_row_key_));
    }
    // left_open
    if (left_open_ && !right_open_) {
      return new BigtableRowRangeResource(
          cbt::RowRange::LeftOpen(left_row_key_, right_row_key_));
    }
    return Status(
        error::INTERNAL,
        "Reached impossible branch. Above clauses should cover all possible "
        "values of left_open_ and right_open_. Please report this issue here:"
        "https://github.com/tensorflow/io/issues/new/choose quoting these"
        "values: left_open:" +
            std::to_string(left_open_) +
            " right_open:" + std::to_string(right_open_));
  }

 private:
  mutex mu_;
  std::string left_row_key_ TF_GUARDED_BY(mu_);
  bool left_open_ TF_GUARDED_BY(mu_);
  std::string right_row_key_ TF_GUARDED_BY(mu_);
  bool right_open_ TF_GUARDED_BY(mu_);
};

REGISTER_KERNEL_BUILDER(Name("BigtableRowRange").Device(DEVICE_CPU),
                        BigtableRowRangeOp);

class BigtablePrintRowRangeOp : public OpKernel {
 public:
  explicit BigtablePrintRowRangeOp(OpKernelConstruction* context)
      : OpKernel(context) {}

  void Compute(OpKernelContext* context) override {
    BigtableRowRangeResource* resource;
    OP_REQUIRES_OK(context,
                   GetResourceFromContext(context, "row_range", &resource));
    core::ScopedUnref unref(resource);

    Tensor* output_tensor = NULL;
    OP_REQUIRES_OK(context, context->allocate_output(0, {1}, &output_tensor));
    auto output_v = output_tensor->tensor<tstring, 1>();

    output_v(0) = resource->ToString();
  }
};

REGISTER_KERNEL_BUILDER(Name("BigtablePrintRowRange").Device(DEVICE_CPU),
                        BigtablePrintRowRangeOp);

class BigtablePrefixRowRangeOp
    : public AbstractBigtableResourceOp<BigtableRowRangeResource> {
 public:
  explicit BigtablePrefixRowRangeOp(OpKernelConstruction* ctx)
      : AbstractBigtableResourceOp<BigtableRowRangeResource>(ctx) {
    VLOG(1) << "BigtablePrefixRowRangeOp ctor ";
    OP_REQUIRES_OK(ctx, ctx->GetAttr("prefix", &prefix_));
  }

 private:
  StatusOr<BigtableRowRangeResource*> CreateResource() override {
    return new BigtableRowRangeResource(cbt::RowRange::Prefix(prefix_));
  }

 private:
  std::string prefix_;
};

REGISTER_KERNEL_BUILDER(Name("BigtablePrefixRowRange").Device(DEVICE_CPU),
                        BigtablePrefixRowRangeOp);

}  // namespace io
}  // namespace tensorflow
